cpp/src/arrow/dataset/large_row_parquet_encrypt_test.cc

```diff
@@ -0,0 +1,201 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <arrow/api.h>
+#include <arrow/dataset/dataset.h>
+#include <arrow/dataset/file_base.h>
+#include <arrow/dataset/file_parquet.h>
+#include <arrow/dataset/parquet_encryption_config.h>
+#include <arrow/filesystem/localfs.h>  // Include the header for LocalFileSystem
+#include <arrow/filesystem/mockfs.h>
+#include <arrow/io/api.h>
+#include <arrow/status.h>
+#include <arrow/table.h>
+#include <arrow/testing/gtest_util.h>
+#include <arrow/testing/random.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/encryption/crypto_factory.h>
+#include <parquet/encryption/encryption.h>
+#include <parquet/encryption/encryption_internal.h>
+#include <parquet/encryption/kms_client.h>
+#include <parquet/encryption/test_in_memory_kms.h>
+#include <cmath>
+#include <memory>
+#include <string_view>
+#include "gtest/gtest.h"
+
+#include <arrow/util/base64.h>
+#include <filesystem>
+
+constexpr int ROW_COUNT = 32769;
+
+constexpr std::string_view kFooterKeyName = "footer_key";
+constexpr std::string_view kColumnKeyMapping = "col_key: foo";
+constexpr std::string_view kBaseDir = "/Users/dtolley/Documents/temp";
+
+using arrow::internal::checked_pointer_cast;
+
+namespace arrow {
+namespace dataset {
+
+class NoOpKmsClient : public parquet::encryption::KmsClient {
+ public:
+  NoOpKmsClient() {}
+
+  std::string WrapKey(const std::string& key_bytes,
+                      const std::string& master_key_identifier) override {
+    // Base64 encode the key
+    return arrow::util::base64_encode(std::string_view(key_bytes));
+  }
+
+  std::string UnwrapKey(const std::string& wrapped_key,
+                        const std::string& master_key_identifier) override {
+    // Base64 decode the key
+    return arrow::util::base64_decode(std::string_view(wrapped_key));
+  }
+};
+
+class NoOpKmsClientFactory : public parquet::encryption::KmsClientFactory {
+ public:
+  NoOpKmsClientFactory() {}
+
+  std::shared_ptr<parquet::encryption::KmsClient> CreateKmsClient(
+      const parquet::encryption::KmsConnectionConfig& config) override {
+    // Return a new instance of NoOpKmsClient
+    return std::make_shared<NoOpKmsClient>();
+  }
+};
+
+class LargeRowEncryptionTest : public ::testing::Test {
```

Comment:
IMO, this test can be simplified and changed to align with the coding style. Do you authorize me to refactor it a little bit?

Reply:
I have tried to refactor it locally with this commit: https://github.com/wgtmac/arrow/commit/93bca5b134ca664bf98853a3d420c39bb6d11005. I can verify that I get following errors occasionally:
```
/Users/gangwu/Projects/arrow/cpp/build_upstream/debug/arrow-dataset-file-parquet-encryption-test --gtest_filter=DatasetEncryptionTest.WriteReadDatasetWithEncryption:DatasetEncryptionTest/*.WriteReadDatasetWithEncryption:DatasetEncryptionTest.WriteReadDatasetWithEncryption/*:*/DatasetEncryptionTest.WriteReadDatasetWithEncryption/*:*/DatasetEncryptionTest/*.WriteReadDatasetWithEncryption:DatasetEncryptionTest.ReadSingleFile:DatasetEncryptionTest/*.ReadSingleFile:DatasetEncryptionTest.ReadSingleFile/*:*/DatasetEncryptionTest.ReadSingleFile/*:*/DatasetEncryptionTest/*.ReadSingleFile:LargeRowEncryptionTest.ReadEncryptLargeRows:LargeRowEncryptionTest/*.ReadEncryptLargeRows:LargeRowEncryptionTest.ReadEncryptLargeRows/*:*/LargeRowEncryptionTest.ReadEncryptLargeRows/*:*/LargeRowEncryptionTest/*.ReadEncryptLargeRows --gtest_color=no
Testing started at 21:32 ...
Running main() from /Users/gangwu/Projects/arrow/cpp/build_upstream/_deps/googletest-src/googletest/src/gtest_main.cc
/Users/gangwu/Projects/arrow/cpp/src/arrow/dataset/file_parquet_encryption_test.cc:154: Failure
Failed
'_error_or_value27.status()' failed with IOError: Failed decryption finalization
/Users/gangwu/Projects/arrow/cpp/src/arrow/dataset/file_parquet_encryption_test.cc:255: Failure
Expected: TestScanDataset() doesn't generate new fatal failures in the current thread.
  Actual: it does.
```

As this test case has many duplication with file_parquet_encryption_test.cc, I simply modified it to incorporate your change. The entire file can be viewed for your convenience: https://[raw.githubusercontent.com/wgtmac/arrow/93bca5b134ca664bf98853a3d420c39bb6d11005/cpp/src/arrow/dataset/file_parquet_encryption_test.cc](https://raw.githubusercontent.com/wgtmac/arrow/93bca5b134ca664bf98853a3d420c39bb6d11005/cpp/src/arrow/dataset/file_parquet_encryption_test.cc)

Reply:
Thanks! I've pushed my local changes of test refactoring to your branch.

Reply:
`cmake .. -G "Unix Makefiles" -DARROW_PARQUET=ON -DARROW_BUILD_TESTS=ON -DARROW_WITH_ZSTD=OFF -DARROW_WITH_ZLIB=OFF -DARROW_WITH_SNAPPY=OFF -DPARQUET_REQUIRE_ENCRYPTION=ON -DARROW_BUILD_BENCHMARKS=OFF -DARROW_DATASET=ON`

That's my cmake command. No special step.

Reply:
Could you try this patch (i.e. without your fix): https://github.com/wgtmac/arrow/commit/93bca5b134ca664bf98853a3d420c39bb6d11005 ?

Reply:
I opened https://github.com/apache/arrow/pull/40256 which is able to reproduce the failure: 

https://github.com/apache/arrow/actions/runs/8058179725/job/22010555480?pr=40256
```
      Start 67: arrow-dataset-file-parquet-encryption-test
    Test #67: arrow-dataset-file-parquet-encryption-test ...***Failed    0.04 sec
Running main() from D:/a/arrow/arrow/build/cpp/_deps/googletest-src/googletest/src/gtest_main.cc
[==========] Running 3 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 2 tests from DatasetEncryptionTest
[ RUN      ] DatasetEncryptionTest.WriteReadDatasetWithEncryption
[       OK ] DatasetEncryptionTest.WriteReadDatasetWithEncryption (8 ms)
[ RUN      ] DatasetEncryptionTest.ReadSingleFile
[       OK ] DatasetEncryptionTest.ReadSingleFile (1 ms)
[----------] 2 tests from DatasetEncryptionTest (10 ms total)

[----------] 1 test from LargeRowEncryptionTest
[ RUN      ] LargeRowEncryptionTest.ReadEncryptLargeRows
D:/a/arrow/arrow/cpp/src/arrow/dataset/file_parquet_encryption_test.cc:154: Failure
Failed
'_error_or_value27.status()' failed with IOError: Failed decryption finalization
D:/a/arrow/arrow/cpp/src/arrow/dataset/file_parquet_encryption_test.cc:255: Failure
Expected: TestScanDataset() doesn't generate new fatal failures in the current thread.
  Actual: it does.
[  FAILED  ] LargeRowEncryptionTest.ReadEncryptLargeRows (3 ms)
[----------] 1 test from LargeRowEncryptionTest (3 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 2 test suites ran. (13 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] LargeRowEncryptionTest.ReadEncryptLargeRows

 1 FAILED TEST
```

Reply:
The test failure is due to race condition so it may not happen at each run.